src/ucs/profile/profile.c

```diff
@@ -226,143 +260,250 @@ static void ucs_profile_write()
+ * @param [in]  file         Source file name.
+ * @param [in]  line         Source line number.
+ * @param [in]  function     Calling function name.
+ * @param [in]  name         Location name.
+ * @param [out] loc_id_p     Filled with location ID:
+ *                             0   - profiling is disabled
+ *                             >0  - location index + 1
  */
-static void ucs_profile_get_location(ucs_profile_type_t type, const char *name,
-                                     const char *file, int line,
-                                     const char *function, volatile int *loc_id_p)
+static UCS_F_NOINLINE
+int ucs_profile_get_location(ucs_profile_type_t type, const char *name,
+                             const char *file, int line, const char *function,
+                             volatile int *loc_id_p)
 {
-    ucs_profile_global_context_t *ctx = &ucs_profile_ctx;
     ucs_profile_global_location_t *loc;
-    int location;
-    int i;
+    int loc_id;
 
-    pthread_mutex_lock(&ucs_profile_ctx.mutex);
+    pthread_mutex_lock(&ucs_profile_global_ctx.mutex);
 
-    if (*loc_id_p == 0) {
+    /* Check, with lock held, that the location is not already initialized */
+    if (*loc_id_p >= 0) {
+        loc_id = *loc_id_p;
         goto out_unlock;
     }
 
     /* Check if profiling is disabled */
     if (!ucs_global_opts.profile_mode) {
-        *loc_id_p = 0;
+        *loc_id_p = loc_id = 0;
         goto out_unlock;
     }
 
     /* Location ID must be uninitialized */
     ucs_assert(*loc_id_p == -1);
 
-    for (i = 0; i < ctx->num_locations; ++i) {
-        loc = &ctx->locations[i];
-
-        if ((type == loc->super.type) &&
-            (line == loc->super.line) &&
+    ucs_profile_for_each_location(loc) {
+        if ((type == loc->super.type) && (line == loc->super.line) &&
             !strcmp(loc->super.name, name) &&
             !strcmp(loc->super.file, basename(file)) &&
             !strcmp(loc->super.function, function)) {
-
-            *loc_id_p = i + 1;
-            goto out_unlock;
+            goto out_found;
         }
     }
 
-    location = ucs_profile_ctx.num_locations++;
+    ++ucs_profile_global_ctx.num_locations;
 
     /* Reallocate array if needed */
-    if (ucs_profile_ctx.num_locations > ucs_profile_ctx.max_locations) {
-        ucs_profile_ctx.max_locations = ucs_profile_ctx.num_locations * 2;
-        ucs_profile_ctx.locations = ucs_realloc(ucs_profile_ctx.locations,
-                                                sizeof(*ucs_profile_ctx.locations) *
-                                                ucs_profile_ctx.max_locations,
-                                                "profile_locations");
-        if (ucs_profile_ctx.locations == NULL) {
+    if (ucs_profile_global_ctx.num_locations > ucs_profile_global_ctx.max_locations) {
+        ucs_profile_global_ctx.max_locations =
+                        2 * ucs_profile_global_ctx.num_locations;
+        ucs_profile_global_ctx.locations =
```

Comment:
pls, use a new temporary variable to store `ucs_realloc()` result
if `ucs_realloc()` returns `NULL`, we lose memory whcih `ucs_profile_global_ctx.locations` points to

Reply:
right